### DISCUSSION TITLE
WaitCursor for Loading Heavy Sets and Updating Viewport

### DISCUSSION THREAD

**Discussion by jallex** [Autodesk/maya-usd#2504] [Q&A]
Hi everyone! We’ve recently tried to add a waitCursor to let the user know that Maya is busy while heavy USD sets are being loaded. We were able to add the waitCursor to the code that loads the set, but once that completes and the waitCursor is deconstructed, Maya still takes quite some time to update the viewport.

We want the waitCursor to stay until the viewport finishes updating, to let the user know Maya isn’t frozen and is still loading their set. Is there something we can hook into to know when the viewport is done updating, or has there been any discussion/ideas on your side about this topic?
Thank you!

**Comment by seando-adsk** [comment]
Have you tried this scenario with a recent version of the MayaUsd plugin compiled from `dev` branch? We recently added a progress bar to many of the operations that are time consuming (such as loading usd file).

**Reply by jallex** [reply]
@seando-adsk Thanks for looking into it! We did try pulling the latest from the dev branch and testing that, but we're not seeing a progress bar or waitCursor when loading Usd sets still.. 
For context, we're loading the sets through right clicking in the outliner context menu and selecting "Load", "Load with Descendents" and "Unload"

**Reply by seando-adsk** [reply]
Ah yes sorry we didn't add the progress bar for those scenarios.

**Reply by gracekumagai** [reply]
Hi @seando-adsk - in this case, do you happen to know if there's a way for us to hook into the viewport update or if you have any suggestions on how we can tackle this on our side? Thank you so much!

**Reply by williamkrick** [reply]
@gracekumagai you can use MUiMessage::add3dViewPostRenderMsgCallback to get a notification when the viewport for a given panel has finished draw the 3d scene for a given panel. 2d elements like the HUD draw after this callback, but that drawing is very fast so for the purpose of turning off the wait cursor it should work.

The default panel used by Maya is usually named "modelPanel4". I know the `getPanel` command can be used to list all the modelPanels in Maya. I don't see a C++ way of listing all the panels so you'd have to call the mel (MGlobal::executeCommand) to know which panels exist to setup the callbacks.

**Reply by gracekumagai** [reply]
Thanks, Will! We will take a look.

**Comment by womanyoyoyo-adsk** [comment]
We have this logged  #1783 and internally as well.
